staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go

```diff
@@ -62,35 +80,50 @@ func (r *ready) wait() error {
 
 // check returns true only if it is Ready.
 func (r *ready) check() bool {
-	// TODO: Make check() function more sophisticated, in particular
-	// allow it to behave as "waitWithTimeout".
-	rwMutex := r.c.L.(*sync.RWMutex)
-	rwMutex.RLock()
-	defer rwMutex.RUnlock()
+	r.lock.RLock()
+	defer r.lock.RUnlock()
 	return r.state == Ready
 }
 
 // set the state to Pending (false) or Ready (true), it does not have effect if the state is Stopped.
 func (r *ready) set(ok bool) {
-	r.c.L.Lock()
-	defer r.c.L.Unlock()
+	r.lock.Lock()
+	defer r.lock.Unlock()
 	if r.state == Stopped {
 		return
 	}
 	if ok {
 		r.state = Ready
-	} else {
+		select {
+		case <-r.waitCh:
+		default:
+			close(r.waitCh)
+		}
+	} else if r.state != Pending {
```

Comment:
We don't want to get back from `Stopped` to Pending.

So it should be changed to:
```
else if r.state == Ready {
```

Reply:
it has already a 
```
	if r.state == Stopped {
		return
	}
```

but as you say this can only enter here if it is Ready so it is better to be explicit